### ISSUE TITLE
[BUG] System.AccessViolationException. GPU Memory Leaks ??

### ISSUE THREAD

**Issue by axel578** [mono/SkiaSharp#1244]
**Description**

I am trying to make some sort of animation on an SkElement in wpf, for that :
First - I create a surface with GrContext set on OpenGL ( For the OpenGL context I Used The GLControl context with the method MakeCurrent()<-- only issue for 64 bits ? ).
Second - I Draw offscreen with this GPU acceleration configuration on a canvas initiated as a Local Variable as much as the surface with the GrContext.
Third - I start InvalidateVisual() for the SkElement every 1/30 seconds, all runs perfectly until at some random( or more some sort of range like 300 ) number of iteration it gives me : System.AccessViolationException: 'Attempted to read or write protected memory. This is often an indication that other memory is corrupt.'

I Tried canvas flush, clear ...
But I Looked up at the GrContext MemoryCache available ( the default limit for it is :100 000 000  )
And I saw that they are some sort of memory leaks on the Gpu side , I made some graphically discussable representation here : 
![image](https://user-images.githubusercontent.com/44734791/79917276-771b6b80-842a-11ea-9e1c-7a8f59548cd3.png)

The White number on the right represent the memory used ( 100 000 000 is the limit by default )
And when I upgrade the memory limit that it can use , the program crashes after way more iterations processed but the problem is still here, they are some strange memory leaks on the GPU side for which I tryed  nearly everything to solve, the only issue is to look up at the memory used and once it's too high, i abandon the gr context and set it up again but i loose a lot of time with this way

And also, when this message "System.AccessViolationException..." appear, it doesn't tell me which line it is
**Code**

```cs
             cnt = new SKGLControl();
            /*glControl.PaintSurface += OnPaintGL;
            glControl.Dock = System.Windows.Forms.DockStyle.Fill;
            glControl.Click += OnSampleClicked;*/
            cnt.Width = 1920;
            cnt.Height = 1080;
            cnt.MakeCurrent();

            interf = GRGlInterface.CreateNativeGlInterface();
            _context = GRContext.Create(GRBackend.OpenGL, interf);
            _context.SetResourceCacheLimits(100, 1000000000);
            // create the surface (UPDATED)
            var glInfo = new GRGlFramebufferInfo(
                fboId: 0,
                format: SKColorType.Rgba8888.ToGlSizedFormat());

            var renderTarget = new GRBackendRenderTarget(
                width: 1920,
                height: 1080,
                sampleCount: 0,
                stencilBits: 0,
                glInfo: glInfo
                );
            

            Surface = SKSurface.Create(
                _context, renderTarget, GRSurfaceOrigin.BottomLeft, SKColorType.Rgba8888);

```

**Expected Behavior**

**Actual Behavior**

**Basic Information**

- Version with issue:  1.68.2 60 preview
- Last known good version:  None
- IDE:  Visual Studio 2019
- Platform Target Frameworks: I personnaly am on 4.7.2

**Comment by mattleibow** [comment]
Is it possible to attach a repro case or point to a repo for this issue.

**Comment by axel578** [comment]
Actually my code is a big big mess ; but the problème is really at this GrContext which used by any Open GL context Creator start to swing liké the image above

**Comment by axel578** [comment]
@mattleibow here is a file which demonstrates the bug: https://drive.google.com/file/d/1NPnwOzxm6ASJy_qXWYioliVM4E7OaErI/view?usp=sharing

**Comment by axel578** [comment]
@mattleibow any issue for this ?

**Comment by mattleibow** [comment]
@axel578 I had a look at the sample, and I think that I got a solution for the issue now building. 

The issue is that the GC is getting a bit desperate to collect things, so it goes ahead and collects the result of `Snapshot`, even though it is being "used" by `ToWriteableBitmap`. The reason this happens is because we don't actually use the image anymore - we use the pixel pointer. The fix just tells the GC that we actually are still using object.

There is an easy fix, and an improvement to the sample code:

```csharp
SuperImageWithGpu.Source = Surface.Snapshot().ToWriteableBitmap();
```

To:
```cs
using (var snapshot = Surface.Snapshot())
	SuperImageWithGpu.Source = snapshot.ToWriteableBitmap();
```

**Comment by mattleibow** [comment]
Fixed by #1258
